File: examples/flutter/core/shared/app_mcp_contracts.dart
import '../domain.case.dart';

class AppMcpClientInfo {
  const AppMcpClientInfo({required this.name, required this.version});

  final String name;
  final String version;

  Dict toJson() {
    return {'name': name, 'version': version};
  }
}

class AppMcpServerInfo {
  const AppMcpServerInfo({
    required this.name,
    required this.version,
    required this.protocolVersion,
    this.instructions,
  });

  final String name;
  final String version;
  final String protocolVersion;
  final String? instructions;
}

class AppMcpInitializeParams {
  const AppMcpInitializeParams({
    required this.protocolVersion,
    this.capabilities,
    this.clientInfo,
  });

  final String protocolVersion;
  final Dict? capabilities;
  final AppMcpClientInfo? clientInfo;
}

class AppMcpInitializeResult {
  const AppMcpInitializeResult({
    required this.protocolVersion,
    required this.capabilities,
    required this.serverInfo,
    this.instructions,
  });

  final String protocolVersion;
  final Dict capabilities;
  final Dict serverInfo;
  final String? instructions;

  Dict toJson() {
    return {
      'protocolVersion': protocolVersion,
      'capabilities': capabilities,
      'serverInfo': serverInfo,
      if (instructions != null) 'instructions': instructions,
    };
  }
}

class AppMcpTextContent {
  const AppMcpTextContent({required this.text}) : type = 'text';

  final String type;
  final String text;

  Dict toJson() {
    return {'type': type, 'text': text};
  }
}

class AppMcpToolDescriptor {
  const AppMcpToolDescriptor({
    required this.name,
    required this.inputSchema,
    this.title,
    this.description,
    this.outputSchema,
    this.annotations,
  });

  final String name;
  final String? title;
  final String? description;
  final AppSchema inputSchema;
  final AppSchema? outputSchema;
  final Dict? annotations;

  Dict toJson() {
    return {
      'name': name,
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      'inputSchema': inputSchema,
      if (outputSchema != null) 'outputSchema': outputSchema,
      if (annotations != null) 'annotations': annotations,
    };
  }
}

class AppMcpResourceDescriptor {
  const AppMcpResourceDescriptor({
    required this.uri,
    required this.name,
    this.title,
    this.description,
    this.mimeType,
    this.annotations,
  });

  final String uri;
  final String name;
  final String? title;
  final String? description;
  final String? mimeType;
  final Dict? annotations;

  Dict toJson() {
    return {
      'uri': uri,
      'name': name,
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      if (mimeType != null) 'mimeType': mimeType,
      if (annotations != null) 'annotations': annotations,
    };
  }
}

class AppMcpTextResourceContent {
  const AppMcpTextResourceContent({
    required this.uri,
    required this.text,
    this.mimeType,
  });

  final String uri;
  final String? mimeType;
  final String text;

  Dict toJson() {
    return {
      'uri': uri,
      if (mimeType != null) 'mimeType': mimeType,
      'text': text,
    };
  }
}

class AppMcpCallResult {
  const AppMcpCallResult({
    required this.content,
    this.structuredContent,
    this.isError,
  });

  final List<AppMcpTextContent> content;
  final dynamic structuredContent;
  final bool? isError;

  Dict toJson() {
    return {
      'content': content.map((item) => item.toJson()).toList(growable: false),
      if (structuredContent != null) 'structuredContent': structuredContent,
      if (isError != null) 'isError': isError,
    };
  }
}

class AppMcpReadResourceResult {
  const AppMcpReadResourceResult({required this.contents});

  final List<AppMcpTextResourceContent> contents;

  Dict toJson() {
    return {
      'contents': contents.map((item) => item.toJson()).toList(growable: false),
    };
  }
}

class AppMcpRequestContext {
  const AppMcpRequestContext({
    required this.transport,
    this.requestId,
    this.sessionId,
    this.correlationId,
    this.clientInfo,
    this.protocolVersion,
  });

  final String transport;
  final dynamic requestId;
  final String? sessionId;
  final String? correlationId;
  final AppMcpClientInfo? clientInfo;
  final String? protocolVersion;
}

abstract interface class AppMcpServer {
  AppMcpServerInfo serverInfo();

  Future<AppMcpInitializeResult> initialize([
    AppMcpInitializeParams? params,
    AppMcpRequestContext? parent,
  ]);

  Future<List<AppMcpToolDescriptor>> listTools([AppMcpRequestContext? parent]);

  Future<List<AppMcpResourceDescriptor>> listResources([
    AppMcpRequestContext? parent,
  ]);

  Future<AppMcpReadResourceResult> readResource(
    String uri, [
    AppMcpRequestContext? parent,
  ]);

  Future<AppMcpCallResult> callTool(
    String name,
    dynamic args, [
    AppMcpRequestContext? parent,
  ]);
}

abstract class BaseAppMcpAdapter {
  String get transport;
}

abstract class BaseAppMcpProcessAdapter extends BaseAppMcpAdapter {
  Future<void> serve(AppMcpServer server);
}

class AppMcpHttpExchange {
  const AppMcpHttpExchange({
    required this.method,
    required this.path,
    required this.headers,
    this.bodyText,
  });

  final String method;
  final String path;
  final Map<String, String?> headers;
  final String? bodyText;
}

class AppMcpHttpResponse {
  const AppMcpHttpResponse({
    required this.statusCode,
    this.headers,
    this.bodyText,
  });

  final int statusCode;
  final Map<String, String>? headers;
  final String? bodyText;
}

abstract class BaseAppMcpHttpAdapter extends BaseAppMcpAdapter {
  String get endpointPath;

  Future<AppMcpHttpResponse?> handle(
    AppMcpHttpExchange exchange,
    AppMcpServer server,
  );
}

class AppMcpProtocolError implements Exception {
  const AppMcpProtocolError(this.code, this.message, [this.data]);

  final int code;
  final String message;
  final dynamic data;

  @override
  String toString() {
    return 'AppMcpProtocolError($code): $message';
  }
}

File: examples/angular/core/shared/app_mcp_contracts.ts
/* ========================================================================== *
 * APP v1.1.1
 * core/shared/app_mcp_contracts.ts
 * ----------------------------------------------------------------------------
 * Protocol-level MCP contracts for APP agent hosts.
 *
 * These contracts define the structural boundary between an agent host runtime
 * and one or more concrete MCP transport adapters. They intentionally stay
 * transport-agnostic:
 * - the host owns catalog, policy, and canonical execution delegation
 * - the provider owns framing, lifecycle wiring, and transport mechanics
 *
 * Concrete MCP implementations belong in registry._providers.
 * When a host exposes multiple transports, it should bind them explicitly as
 * named providers such as _providers.mcpAdapters.stdio and
 * _providers.mcpAdapters.http.
 * ========================================================================== */

import type { AppSchema, Dict } from "../domain.case";

export interface AppMcpClientInfo {
  name: string;
  version: string;
}

export interface AppMcpServerInfo {
  name: string;
  version: string;
  protocolVersion: string;
  instructions?: string;
}

export interface AppMcpInitializeParams {
  protocolVersion: string;
  capabilities?: Dict;
  clientInfo?: AppMcpClientInfo;
}

export interface AppMcpInitializeResult {
  protocolVersion: string;
  capabilities: {
    tools?: {
      listChanged?: boolean;
    };
    resources?: {
      listChanged?: boolean;
    };
  };
  serverInfo: {
    name: string;
    version: string;
  };
  instructions?: string;
}

export interface AppMcpTextContent {
  type: "text";
  text: string;
}

export interface AppMcpToolDescriptor {
  name: string;
  title?: string;
  description?: string;
  inputSchema: AppSchema;
  outputSchema?: AppSchema;
  annotations?: Dict;
}

export interface AppMcpResourceDescriptor {
  uri: string;
  name: string;
  title?: string;
  description?: string;
  mimeType?: string;
  annotations?: Dict;
}

export interface AppMcpTextResourceContent {
  uri: string;
  mimeType?: string;
  text: string;
}

export interface AppMcpCallResult {
  content: AppMcpTextContent[];
  structuredContent?: unknown;
  isError?: boolean;
}

export interface AppMcpReadResourceResult {
  contents: AppMcpTextResourceContent[];
}

export interface AppMcpRequestContext {
  transport: string;
  requestId?: string | number | null;
  sessionId?: string;
  correlationId?: string;
  clientInfo?: AppMcpClientInfo;
  protocolVersion?: string;
}

export interface AppMcpServer {
  serverInfo(): AppMcpServerInfo;
  initialize(
    params?: AppMcpInitializeParams,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpInitializeResult>;
  listTools(
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpToolDescriptor[]>;
  listResources(
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpResourceDescriptor[]>;
  readResource(
    uri: string,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpReadResourceResult>;
  callTool(
    name: string,
    args: unknown,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpCallResult>;
}

export abstract class BaseAppMcpAdapter {
  public abstract readonly transport: string;
}

export abstract class BaseAppMcpProcessAdapter extends BaseAppMcpAdapter {
  public abstract serve(server: AppMcpServer): Promise<void>;
}

export interface AppMcpHttpExchange {
  method: string;
  path: string;
  headers: Record<string, string | undefined>;
  bodyText?: string;
}

export interface AppMcpHttpResponse {
  statusCode: number;
  headers?: Record<string, string>;
  bodyText?: string;
}

export abstract class BaseAppMcpHttpAdapter extends BaseAppMcpAdapter {
  public abstract readonly endpointPath: string;
  public abstract handle(
    exchange: AppMcpHttpExchange,
    server: AppMcpServer
  ): Promise<AppMcpHttpResponse | undefined>;
}

export class AppMcpProtocolError extends Error {
  public readonly code: number;
  public readonly data?: unknown;

  constructor(code: number, message: string, data?: unknown) {
    super(message);
    this.name = "AppMcpProtocolError";
    this.code = code;
    this.data = data;
  }
}
